tiatoolbox/models/engine/nucleus_instance_segmentor.py

```diff
@@ -0,0 +1,704 @@
+            if tile_flag[idx] or tile_mode == 3
+        ]
+        sel_indices = [
+            tile_rtree.query(bounds, predicate="contains") for bounds in sel_boxes
+        ]
+    elif tile_mode in [1, 2]:
+        # for `horizontal/vertical strip` tiles
+        # -- extend from the marked edges (top/bot or left/right) by
+        #    the margin size, remove all nuclei lie within the margin
+        #    area (including on the margin line)
+        # -- remove all nuclei on the boundary also
+
+        sel_boxes = [
+            margin_boxes[idx] if flag else boundary_lines[idx]
+            for idx, flag in enumerate(tile_flag)
+        ]
+        sel_indices = [
+            tile_rtree.query(bounds, predicate="intersects") for bounds in sel_boxes
+        ]
+    else:
+        raise ValueError(f"Unknown tile mode {tile_mode}.")
+
+    def retrieve_sel_uids(sel_indices, inst_dict):
+        """Helper to retrieved selected instance uids."""
+        sel_uids = []
+        if len(sel_indices) > 0:
+            # not sure how costly this is in large dict
+            inst_uids = list(inst_dict.keys())
+            sel_indices = [idx for sub_sel in sel_indices for idx in sub_sel]
+            sel_uids = [inst_uids[idx] for idx in sel_indices]
+        return sel_uids
+
+    remove_insts_in_tile = retrieve_sel_uids(sel_indices, inst_dict)
+
+    # external removal only for tile at cross sections
+    # this one should contain UUID with the reference database
+    remove_insts_in_orig = []
+    if tile_mode == 3:
+        # ! DEPRECATION:
+        # !     will be deprecated upon finalization of SQL annotation store
+        inst_boxes = [v["box"] for v in ref_inst_dict.values()]
+        inst_boxes = np.array(inst_boxes)
+        ref_inst_rtree = pygeos.STRtree(
+            pygeos.box(
+                inst_boxes[:, 0],
+                inst_boxes[:, 1],
+                inst_boxes[:, 2],
+                inst_boxes[:, 3],
+            )
+        )
+        # !
+
+        # remove existing instances in old prediction which intersect
+        # with the margin lines
+        sel_indices = [
+            ref_inst_rtree.query(bounds, predicate="intersects")
+            for bounds in margin_lines
+        ]
+        remove_insts_in_orig = retrieve_sel_uids(sel_indices, ref_inst_dict)
+
+    # move inst position from Tile space back to WSI space
+    # an also generate universal uid as replacement for storage
+    new_inst_dict = {}
+    for inst_uid, inst_info in inst_dict.items():
+        if inst_uid not in remove_insts_in_tile:
+            inst_info["box"] += np.concatenate([tile_tl] * 2)
+            inst_info["centroid"] += tile_tl
+            inst_info["contour"] += tile_tl
+            inst_uuid = uuid.uuid4().hex
+            new_inst_dict[inst_uuid] = inst_info
+
+    return new_inst_dict, remove_insts_in_orig
+
+
+class NucleusInstanceSegmentor(SemanticSegmentor):
+    """An engine specifically designed to handle WSI inference.
```

Comment:
What about tiles ? We need to be able to apply functionality for that too as it will be very commonly required